text/3738-drop-type-destructuring.md

```diff
@@ -0,0 +1,392 @@
+    // `a` is moved out of `x`, Drop never runs for `x`
+    destructure!(let Foo { a } = x);
+}
+```
+
+Preventing `Drop` from running is not inherently problematic in Rust. You can already do this for any type through `mem::forget` (and other means), which is safe.
+
+
+Finally, the following is an example of how one could implement the example from [Motivation](#motivation):
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    let mut this = ManuallyDrop::new(self);
+    let buf = mem::take(&mut this.buf);
+    let buf = if !this.panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    // SAFETY: double-drops are prevented by putting `this` in a ManuallyDrop that is never dropped
+    let inner = unsafe { ptr::read(&this.inner) };
+
+    (inner, buf)
+}
+```
+
+Now with `destructure!`:
+
+```rust
+pub fn into_parts(self) -> (W, Result<Vec<u8>, WriterPanicked>) {
+    destructure!(let Self { buf, inner, panicked } = self);
+    let buf = if !panicked { Ok(buf) } else { Err(WriterPanicked { buf }) };
+
+    (inner, buf)
+}
+```
+
+We don't even need unsafe anymore :tada:
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A built-in macro with the following "signature" is added to the standard library:
+
+```rust
+#[macro_export]
+macro_rules! destructure {
+    ((let)? $pat:pat = $e:expr) => { /* built-in */ }
+}
+```
+
+The result of the expansion checks the following requirements and produces a compilation error if any of them are not met:
+- `$pat` must be an irrefutable pattern
+- `$e` must be an owned expression that can be matched by pattern `$pat`
+- `$e`'s type must be an ADT (`struct` or `enum`)
+- The type of `$e` must be constructable in the current context, i.e.
+    - All fields must be visible to the current module wrt privacy
+    - If the type is marked with `#[non_exhaustive]`, it must be defined in the current crate
+
+The semantic of this macro is equivalent to the following:
+1. Wrapping `$e` in a `ManuallyDrop`
+2. Copying **all** fields of `$e` (for enums all fields of the active variant)
+3. If the `let` token is present in the macro invocation, fields that are listed in `$pat` are binded to their respective bindings or patterns. Otherwise, they are assigned similarly to destructuring assignment
+4. If `..` is present in `$pat`, all fields that are not mentioned in it are dropped in the order of definition (simialrly to how [normal `let`/destructuring assignment works](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3abd3aebd3378690ff3d2006e12d4120))
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+When implementing this change as a macro expanding to a statement, we do not think there are many drawbacks. However, in the coming section with alternatives we discuss another possible expansion to a pattern which may have a few advantages, but also has many drawbacks which we discuss there.
+
+# Rationale and alternatives
```

Comment:
Another alternative that I haven't seen mentioned is having a macro that behaves like a match. Something like:

```rust
match_and_forget!(foo { Foo { .. } => something)
```

The main benefit, is that it would be straightforward to get that to work for enums:

```rust
match_and_forget!( thing {
    Variant1(x) => something(x),
    Variant2{ y, z } => something_else(y, z),
});
```

And it also would allow using it as an expression.

The downside is that it is a lot more verbose for simply binding fields to variables for a struct.

Reply:
I think this is basically just equivalent to the `destructure!` as a pattern, but more verbose for simple cases. It also suffers from similar inability to do nested matches.

Also:
> The one scenario we could imagine is that might be useful for `destructure!()` on `enum`s, but we also think that even using `destructure!()` on `enum`s will be very uncommon (as types implementing `Drop` are almost always `struct`s) and might not outweigh the increased implementation complexity and the fact that we need to communicate this one level matching rule to users.